adapters/lora/src/main/java/org/eclipse/hono/adapter/lora/LoraCommandSubscriptions.java

```diff
@@ -0,0 +1,175 @@
+                    }
+                });
+
+        NotificationEventBusSupport.registerConsumer(vertx, DeviceChangeNotification.TYPE,
+                notification -> {
+                    if (LifecycleChange.DELETE.equals(notification.getChange())) {
+                        closeConsumersForGateway(notification.getTenantId(), notification.getDeviceId());
+                    }
+                });
+    }
+
+    @SuppressWarnings("rawtypes")
+    private void closeConsumersForGateway(final String tenantId, final String gatewayId) {
+        final SubscriptionKey key = new SubscriptionKey(tenantId, gatewayId);
+        final TriTuple<CommandConsumer, LoraProvider, Context> subscription = commandSubscriptions.remove(key);
+        if (subscription == null) {
+            return;
+        }
+
+        final Span span = TracingHelper
+                .buildSpan(tracer, null, "close command subscription for deleted gateway",
+                        getClass().getSimpleName())
+                .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_CLIENT)
+                .start();
+        TracingHelper.setDeviceTags(span, tenantId, gatewayId);
+        final List<Future> consumerCloseFutures = Collections.synchronizedList(new ArrayList<>());
+        final Context ctx = subscription.three();
+        ctx.runOnContext(x -> {
+            final CommandConsumer commandConsumer = subscription.one();
+            consumerCloseFutures.add(commandConsumer.close(null));
+        });
+        CompositeFuture.join(consumerCloseFutures).onComplete(x -> span.finish());
+    }
+
+    @SuppressWarnings("rawtypes")
+    private void closeConsumersForTenant(final String tenantId) {
+        if (!commandSubscriptions.entrySet().stream().anyMatch(e -> e.getKey().getTenant().equals(tenantId))) {
+            return;
+        }
+
+        final Span span = TracingHelper
+                .buildSpan(tracer, null, "close command subscriptions for disabled/deleted tenant",
+                        getClass().getSimpleName())
+                .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_CLIENT)
+                .start();
+        TracingHelper.setDeviceTags(span, tenantId, null);
+        final List<Future> consumerCloseFutures = Collections.synchronizedList(new ArrayList<>());
+        final var iter = commandSubscriptions.entrySet().stream()
+                .filter(e -> e.getKey().getTenant().equals(tenantId)).iterator();
+        while (iter.hasNext()) {
+            final var subscription = iter.next();
+            final Context ctx = subscription.getValue().three();
+            ctx.runOnContext(x -> {
+                final CommandConsumer commandConsumer = subscription.getValue().one();
+                consumerCloseFutures.add(commandConsumer.close(null));
+            });
+            iter.remove();
+        }
+        CompositeFuture.join(consumerCloseFutures).onComplete(x -> span.finish());
```

Comment:
I think it is probably better to finish the span here on the same thread/vert.x-context as the one on which the span was created (using `runOnContext`).